stan/math/prim/fun/read_cov_matrix.hpp

```diff
@@ -13,36 +13,42 @@ namespace math {
  * A generally worse alternative to call prior to evaluating the
  * density of an elliptical distribution
  *
- * @tparam T type of elements in the arrays
+ * @tparam T_CPCs type of \c T_CPCs
+ * @tparam T_sds type of \c sds
  * @param CPCs on (-1, 1)
  * @param sds on (0, inf)
  * @param log_prob the log probability value to increment with the Jacobian
  * @return Covariance matrix for specified partial correlations.
  */
-template <typename T>
-Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> read_cov_matrix(
-    const Eigen::Array<T, Eigen::Dynamic, 1>& CPCs,
-    const Eigen::Array<T, Eigen::Dynamic, 1>& sds, T& log_prob) {
-  Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> L
+template <typename T_CPCs, typename T_sds,
+          require_all_eigen_vector_t<T_CPCs, T_sds>* = nullptr,
+          require_same_vt<T_CPCs, T_sds>* = nullptr>
+Eigen::Matrix<value_type_t<T_CPCs>, Eigen::Dynamic, Eigen::Dynamic>
+read_cov_matrix(const T_CPCs& CPCs, const T_sds& sds,
```

Comment:
This feels like a good candidate for forwarding

Reply:
I see no need for perfect forwarding here.

Reply:
Why not? It looks like they are just being forwarded along to read_cov_L?

I started trying to write a test to look at const& vs. &&, what do you think about the below?

https://github.com/seantalts/perf-math/blob/test/eigen_pf/eigen_pf.cpp

I used `noinline` to act like a larger function that would not be an inline candidate and no_icf because I think preserving the original functions is a thing that makes sense (mby idk?)

Besides that it just does a mul passing the values down two layers of functions. You should be able to compile and run it with

```
make eigen_pf && ./eigen_pf
```

I'm running a longer test ala
```
./eigen_pf --benchmark_repetitions=50 --benchmark_report_aggregates_only=true
```

To get a better idea of how the large data benchmarks run on average

Reply:
One run on mine says it's about 1-2% faster as a ballpark, though I'd rather wait for the numbers from a few repetitions to get an idea of pf is useful in these situations

```
Running ./eigen_pf
Run on (32 X 3500 MHz CPU s)
CPU Caches:
  L1 Data 32K (x16)
  L1 Instruction 64K (x16)
  L2 Unified 512K (x16)
  L3 Unified 8192K (x4)
Load Average: 1.44, 1.60, 1.41
--------------------------------------------------------------
Benchmark                    Time             CPU   Iterations
--------------------------------------------------------------
add_copy_bench/2           176 ns          176 ns      4040759
add_copy_bench/4           320 ns          320 ns      2192642
add_copy_bench/8          1112 ns         1112 ns       626119
add_copy_bench/16         4275 ns         4274 ns       163690
add_copy_bench/32        17538 ns        17537 ns        39809
add_copy_bench/64        78332 ns        78331 ns         8988
add_copy_bench/128      412057 ns       412041 ns         1702
add_copy_bench/256     2411324 ns      2411265 ns          298
add_copy_bench/512    14329605 ns     14329181 ns           48
add_copy_bench/1024   92212267 ns     92210599 ns            7
add_copy_bench/2048  667639201 ns    667627868 ns            1
add_copy_bench/4096 5055086122 ns   5054647216 ns            1
add_copy_bench/8192 38807557827 ns   38803388581 ns            1
add_pf_bench/2             175 ns          175 ns      3985378
add_pf_bench/4             318 ns          318 ns      2196480
add_pf_bench/8            1161 ns         1161 ns       611528
add_pf_bench/16           4424 ns         4424 ns       157600
add_pf_bench/32          18263 ns        18262 ns        38474
add_pf_bench/64          81782 ns        81761 ns         8585
add_pf_bench/128        388528 ns       388511 ns         1791
add_pf_bench/256       2120808 ns      2120470 ns          335
add_pf_bench/512      13709277 ns     13707653 ns           49
add_pf_bench/1024     89777511 ns     89767389 ns            7
add_pf_bench/2048    701121667 ns    701069035 ns            1
add_pf_bench/4096   4939446339 ns   4939085649 ns            1
add_pf_bench/8192   38407552743 ns   38404496372 ns            1
```

Reply:
Benchmark code seems fine for testing passing in rvalues. For completness maybe you should also test lvalues. I ran 50 iterations over a bit sparser range of sizes and it seems pf is a bit than passing by const ref. Makes sense, since pf needs to invoke move constructor and alternative just passes pointer. Running the benchmark a few more times differencer vary. I think it is safe to say these differences are most likely just noise.

```
Running F:\Users\tadej\Programming\bstatcomp\perf-math\eigen_pf.exe
Run on (4 X 3293 MHz CPU s)
CPU Caches:
  L1 Data 32K (x4)
  L1 Instruction 32K (x4)
  L2 Unified 262K (x4)
  L3 Unified 6291K (x1)
------------------------------------------------------------------
Benchmark                           Time           CPU Iterations
------------------------------------------------------------------
add_copy_bench/2_mean             453 ns        452 ns    1544828
add_copy_bench/2_median           452 ns        455 ns    1544828
add_copy_bench/2_stddev             1 ns          5 ns    1544828
add_copy_bench/4_mean             832 ns        831 ns     896000
add_copy_bench/4_median           831 ns        837 ns     896000
add_copy_bench/4_stddev             5 ns          9 ns     896000
add_copy_bench/16_mean           8900 ns       8862 ns      89600
add_copy_bench/16_median         8859 ns       8894 ns      89600
add_copy_bench/16_stddev          123 ns         91 ns      89600
add_copy_bench/64_mean         156052 ns     155538 ns       3733
add_copy_bench/64_median       154832 ns     154869 ns       3733
add_copy_bench/64_stddev         2452 ns       2294 ns       3733
add_copy_bench/256_mean       3902117 ns    3877444 ns        179
add_copy_bench/256_median     3873370 ns    3840782 ns        179
add_copy_bench/256_stddev       63451 ns      53170 ns        179
add_copy_bench/1024_mean    137456676 ns  137062500 ns          5
add_copy_bench/1024_median  136603340 ns  137500000 ns          5
add_copy_bench/1024_stddev    1902190 ns    1895088 ns          5
add_copy_bench/2048_mean    941109244 ns  936875000 ns          1
add_copy_bench/2048_median  934882400 ns  937500000 ns          1
add_copy_bench/2048_stddev   13821822 ns   12611113 ns          1
add_pf_bench/2_mean               458 ns        456 ns    1544828
add_pf_bench/2_median             454 ns        455 ns    1544828
add_pf_bench/2_stddev               7 ns          7 ns    1544828
add_pf_bench/4_mean               838 ns        834 ns     896000
add_pf_bench/4_median             830 ns        837 ns     896000
add_pf_bench/4_stddev              14 ns         16 ns     896000
add_pf_bench/16_mean             8931 ns       8873 ns      74667
add_pf_bench/16_median           8862 ns       8789 ns      74667
add_pf_bench/16_stddev            129 ns        164 ns      74667
add_pf_bench/64_mean           155515 ns     155064 ns       4480
add_pf_bench/64_median         154515 ns     153460 ns       4480
add_pf_bench/64_stddev           2200 ns       2462 ns       4480
add_pf_bench/256_mean         3894863 ns    3877005 ns        187
add_pf_bench/256_median       3867772 ns    3843583 ns        187
add_pf_bench/256_stddev         62613 ns      53382 ns        187
add_pf_bench/1024_mean      138726830 ns  137375000 ns          5
add_pf_bench/1024_median    138238250 ns  137500000 ns          5
add_pf_bench/1024_stddev      1994051 ns    2358901 ns          5
add_pf_bench/2048_mean      948947858 ns  941562500 ns          1
add_pf_bench/2048_median    942359850 ns  937500000 ns          1
add_pf_bench/2048_stddev     22353437 ns   16345018 ns          1
```

Reply:
> Why not? It looks like they are just being forwarded along to read_cov_L?

In this case using pf does not avoid any copying. It is just more code.

Reply:
I ran this for 50 iterations for each (data [here](https://gist.github.com/SteveBronder/5d0c064f3b951b3dcd554601b660ee11)) and it looks like moves were pretty beneficial!

![image](https://user-images.githubusercontent.com/5857231/75798746-04a0e000-5d45-11ea-88bc-410f29d47b7f.png)


I can't explain the big spike at around 1000, I actually ran this twice and saw that same spike both times. Maybe something to do with my CPU?

Also fyi I consider this a side convo it doesn't have to be resolved for the PR to go through

Reply:
> In this case using pf does not avoid any copying. It is just more code.

It's not that pf avoids copying it's that pf transfers ownership of the memory and avoids aliasing issues. See the comment below

https://github.com/stan-dev/math/pull/1720#discussion_r381542520

Reply:
Bob asked that I close these and move them over to discourse so I'm gonna post over there and close this